Add quicktip repeated queries (#1065)
getkirby/getkirby.com: content/1_docs/2_cookbook/3_templating/0_keep_your_code_dry/recipe.txt

Title: Quicktip: Don't repeat the same query multiple times

----

Published: 2021-01-09

----

> Description: Don't repeat yourself is a coding paradigm that when not heeded can seriously harm performance.

Reviewer: ```suggestion
Description: "Don't repeat yourself" (DRY) is a coding paradigm that when not heeded can seriously harm performance.
```

However, what is true for the big picture, is also true for the smaller chunks of code we produce daily in our templates.

Here is a simple example of what happens quite often:

We define a variable in our controller and pass it to the template:

```php "/site/controllers/blog.php"
<?php

return function($page) {
  $articles = $page->articles()->listed();
  return [
    'articles' => $articles,
  ];
};
```

And then in the template:

```php "/site/temlates/blog.php"
<?php
foreach($page->articles()->listed() as $article) {
  echo $article->title();
}
?>
```

What's wrong here? And why is this a problem? Although we defined the variable in the controller, we do not use it in the template, but perform the same query again as before in the controller.

This means that Kirby has to perform the same query twice instead of using the result stored in the `$articles` variable.

Furthermore, this often results in unexpected behaviour. Consider this example:

```php "/site/controllers/blog.php"
<?php

return function($page) {
  $articles = $page->articles()->listed();
  if($tag = param('tag')) {

>     $articles = $articles->filterBy('tags', $tag, ',');

Reviewer: ```suggestion
    $articles = $articles->filter('tags', $tag, ',');
```

  }
  return [
    'articles' => $articles,
  ];
};
```

> If we now use the same template code as in the example above, we will probably wonder why the filter doesn't work. So if this happens, check if you haven't made a new unfiltered query in the template.

Reviewer: ```suggestion
If we now use the same template code as in the example above, we will probably be surprised why the filter doesn't work. So if this happens, check if you haven't made a new unfiltered query in the template.
```

We should therefore fix this and store the result of the queries in variables first, then replace the repeated queries with these variables:

```php
<?php if (!isset($subpages)) $subpages = $site->children(); ?>

<?php foreach ($subpages->listed() as $p) : ?>
  <?php
  $depth             = $p->depth();
  $hasListedChildren = $p->hasListedChildren()
  ?>
  <li class="depth-<?= $depth ?><?= $depth === 1 ? ' nav-item ' . $p->pageRootClass() : '' ?>
    <?= $hasListedChildren && $depth < 2 ? ' dropdown ' : '' ?>
    <?= $hasListedChildren ? ' has-dimmer ' : '' ?>">
    <a class="nav-link<?= $depth > 1 ? ' dropdown-item ' : '' ?>
    <?= $p->isActive() ? ' active' : '' ?>
    <?= $hasListedChildren && !$depth < 2 ? ' dropdown-toggle ' : ' icon-arrow ' ?>" href="<?= $p->url() ?>">
      <?= $p->title() ?>
    </a>
    <?php if ($hasListedChildren && $depth < 3) : ?>
      <ul id="dropdown-<?= md5($p->url()) ?>" class="dropdown-menu<?= $hasListedChildren && $depth >= 2 ? ' submenu ' : '' ?>">
        <?php snippet('treemenu', ['subpages' => $p->children()]); ?>
      </ul>
    <?php endif ?>
  </li>
<?php endforeach ?>
```

## Takeaway

> Always store query results in variables when they are used multiple times. This improves your overall performance, keeps your code dry and easier to read, and prevents unexpected results.

Reviewer: ```suggestion
Always store query results in variables when they are used multiple times. This improves your overall performance, keeps your code DRY and easier to read, and prevents unexpected results.
```
